Add React Testing Part One Lesson (#23222)
TheOdinProject/curriculum: javascript/react-js/react-testing-part-one.md

### Introduction

We've learned testing our vanilla JavaScript applications in the previous section. Testing is indeed a powerful tool that allows to write maintainable and flexible code. If you've followed along our lessons, we've been using the [Jest](https://jestjs.io/) framework. For React, we'll keep using Jest and add more capabilities to our tests using the React Testing library.

> Read the [introduction](https://testing-library.com/docs/react-testing-library/intro) page of the library before we get started.

Author: No particular reason, just thought it would be a good read to kick off the lesson

> Read the [introduction](https://testing-library.com/docs/react-testing-library/intro) page of the library before we get started.

Reviewer: I think that it would be better for people to just dive into the lesson, but if you feel fairly strongly that this article provides some good benefits before really digging into things I wouldn't call this a requirement for me to approve 🙂

> Read the [introduction](https://testing-library.com/docs/react-testing-library/intro) page of the library before we get started.

Author: Considering that I don't have a strong reason, I'll go ahead and remove it

### Learning Outcomes

By the end of this lesson, you should be able to do the following:

TODO

### Setting Up

We'll need to import some packages atop our test file:

~~~javascript
import React from "react"
import { ... } from "@testing-library/react"
import "@testing-library/jest-dom"  // optional
import userEvent from "@testing-library/user-event" // optional
import TestComponent from "path-to-test-component"
~~~

* `@testing-library/react` will give us access to useful functions like `render` which we'll demonstrate later on.

* `@testing-library/jest-dom` includes some handy custom matchers (assertive functions) like `toBeInTheDocument` and more. (complete list [here](https://github.com/testing-library/jest-dom)). Jest already has a lot of matchers so this package is not compulsory to use. 

> * `@testing-library/user-event` provides the `userEvent` API that simulates user interactions with the webpage. It's optional because we can import the `fireEvent` API from the `@testing-library/react` instead; `userEvent` builds off of `fireEvent` to provide extra features. 

Author: Yes, I agree with your sentiment. I've updated the lesson to prefer `userEvent` over `fireEvent`

### Our First Query

First, we'll render the component using `render`. The API will return an object and we'll use destructuring syntax to obtain a subset of the methods required. You can read all about what `render` can do [here](https://testing-library.com/docs/react-testing-library/api/#render).

~~~javascript
// App.js

import React from "react"

const App = () => <h1>Our First Test</h1>

export default App
~~~

~~~javascript
// App.test.js

import React from "react"
import { render } from "@testing-library/react"
import App from "./App"

test("correct heading renders", () => {
    const { getByRole } = render(<App />)
    expect(getByRole("heading").textContent).toMatch(/our first test/i)
})
~~~

Execute `npm test App.js` on the terminal and you'll see that test pass. `getByRole` is just one of the dozen query methods that we could've used. Essentially, queries are classified into three types: `getBy`, `queryBy` and `findBy`. Read about their differences [here](https://testing-library.com/docs/queries/about/).  

> 

Reviewer: I think that it would be beneficial to touch on using the name option with getByRole, as in my experience people are often unaware of that functionality and that it's a "better" and "more RTL" approach to getting an element than using getByText or a test  ID.

Excerpt from [the RTL docs](https://testing-library.com/docs/queries/byrole/) advocating for the use of the name option with getByRole: 
![image](https://user-images.githubusercontent.com/7115053/149677279-a69ff7c0-7eea-4e0a-8097-d426e988f2db.png)

Reviewer: I'd also like to see us explicitly talk about the RTL approach/ethos when it comes to [which queries should be favored over others](https://testing-library.com/docs/queries/about/#priority) in general

~~~javascript
// App.test.js.snap

// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`magnificent monkeys render 1`] = `
<div>
  <button
    type="button"
  >
    Click Me
  </button>
  <h1>
    Magnificent Monkeys
  </h1>
</div>
`;
~~~

It's an HTML representation of the `App` component. And it will be compared against the `App` in future snapshot assertions. If the `App` changes even slightly, the test fails. 

Snapshot tests are fast and easy to write. One assertion saves us from writing multiple lines of code. For example, with a `toMatchSnapshot`, we're spared of asserting the existence of the button and the heading. They also don't let unexpected changes creep into our code. Read all about what can be achieved with snapshots [here](https://jestjs.io/docs/snapshot-testing)

Snapshots might seem the best thing that has happened to us while testing thus far. But we are forced to wonder, _what_ exactly are we testing? What's being validated? If a snapshot passes, what does it convey about the correctness of the component? 

> The other issue with snapshots is false negatives. Even the insignificant of changes compels the test to fail. Fixing punctuation? Snapshot will fail. Replacing an HTML tag to a more semantic one? Snapshot will fail. This might cause us to lose our confidence in the test suite altogether. Snapshots aren't inherently bad; they do serve a purpose. But it's beneficial to understand when to snapshot, and when not to snapshot. 

Reviewer: ```suggestion
The other issue with snapshots is false negatives. Even the most insignificant of changes compel the test to fail. Fixing punctuation? Snapshot will fail. Replacing an HTML tag to a more semantic one? Snapshot will fail. This might cause us to lose our confidence in the test suite altogether. Snapshots aren't inherently bad; they do serve a purpose. But it's beneficial to understand when to snapshot, and when not to snapshot. 
```

> The other issue with snapshots is false negatives. Even the insignificant of changes compels the test to fail. Fixing punctuation? Snapshot will fail. Replacing an HTML tag to a more semantic one? Snapshot will fail. This might cause us to lose our confidence in the test suite altogether. Snapshots aren't inherently bad; they do serve a purpose. But it's beneficial to understand when to snapshot, and when not to snapshot. 

Reviewer: I would also mention that overreliance on snapshot tests can cause false *positives* and make developers more confident about their code than they should be.

### Assignment

<div class="lesson-content__panel" markdown="1">

> 1. Take a glance at all of the available [query methods](https://testing-library.com/docs/dom-testing-library/cheatsheet/). There's no need to use them all. But it's optimal to employ a specific method for a specific query. If none of the query methods suffice, there's an option to use [test ids](https://testing-library.com/docs/queries/bytestid/).

Reviewer: ```suggestion
1. Take a glance at all of the available [query methods](https://testing-library.com/docs/dom-testing-library/cheatsheet/). There's no need to use them all, but it's optimal to employ a specific method for a specific query. If none of the query methods suffice, there's an option to use [test ids](https://testing-library.com/docs/queries/bytestid/).
```

> 2. Read [fireEvent](https://testing-library.com/docs/dom-testing-library/api-events/) and [userEvent](https://testing-library.com/docs/ecosystem-user-event/) API docs to get a feel of how to achieve user simulation. 

Reviewer: See previous comments regarding fireEvent v userEvent. IMO we don't need to even be pointing people to look at fireEvent.

3. This [article](https://tsh.io/blog/pros-and-cons-of-jest-snapshot-tests/) goes in depth regarding the pros and cons of snapshot testing. And [this](https://www.sitepen.com/blog/snapshot-testing-benefits-and-drawbacks) one does an excellent job of explaining what is snapshot testing in programming in general.

</div>

### Additional Resources

This section contains helpful links to other content. It isn't required, so consider it supplemental.

- This [tutorial](https://academind.com/tutorials/testing-react-apps) is a great overview of what you've learned. It goes into testing async code and callbacks which we haven't covered yet. Though you should be able to follow along using your previous knowledge.

- [This video](https://www.youtube.com/watch?v=YQLn7ycfzEo) for a hands-on tutorial.

### Knowledge Check 

This section contains questions for you to check your understanding of this lesson. If you're having trouble answering the questions below on your own, review the material above to find the answer.

> TODO

Reviewer: Planning on doing these?

import userEvent from "@testing-library/user-event";
import TestComponent from "path-to-test-component";
~~~

* `@testing-library/react` will give us access to useful functions like `render` which we'll demonstrate later on.

* `@testing-library/jest-dom` includes some handy custom matchers (assertive functions) like `toBeInTheDocument` and more. (complete list on [jest-dom's github](https://github.com/testing-library/jest-dom)). Jest already has a lot of matchers so this package is not compulsory to use. 

* `@testing-library/user-event` provides the `userEvent` API that simulates user interactions with the webpage. Alternatively, we could import the `fireEvent` API from `@testing-library/react`. 

  > _Note: `fireEvent` is an inferior counterpart to `userEvent` and `userEvent` should always be preferred in practice._

* No need to import `jest` since it will automatically detect test files (`*.test.js` or `*.test.jsx`).

That's a lot of setup. But good news! If you're initializing your React repositories with `create-react-app`, then all the above packages come preinstalled and the scripts preconfigured in `package.json`. 

### Our First Query

> First, we'll render the component using `render`. The API will return an object and we'll use destructuring syntax to obtain a subset of the methods required. You can read all about what `render` can do in the [render docs](https://testing-library.com/docs/react-testing-library/api/#render).

Reviewer: ```suggestion
First, we'll render the component using `render`. The API will return an object and we'll use destructuring syntax to obtain a subset of the methods required. You can read all about what `render` can do in [the React Testing Library API docs about render](https://testing-library.com/docs/react-testing-library/api/#render).
```

~~~javascript
// App.js

import React from "react";

const App = () => <h1>Our First Test</h1>;

export default App;
~~~

~~~javascript
// App.test.js

import React from "react";
import { render } from "@testing-library/react";
import App from "./App";

test("correct heading renders", () => {
    const { getByRole } = render(<App />);
    expect(getByRole("heading").textContent).toMatch(/our first test/i);
});
~~~

> Execute `npm test App.js` on the terminal and you'll see that test pass. `getByRole` is just one of the dozen query methods that we could've used. Essentially, queries are classified into three types: `getBy`, `queryBy` and `findBy`. Go through the [about queries](https://testing-library.com/docs/queries/about/) page. Pay extra attention to the "Types of Queries" and "Priority" section. 

Reviewer: ```suggestion
Execute `npm test App.js` on the terminal and you'll see that test pass. `getByRole` is just one of the dozen query methods that we could've used. Essentially, queries are classified into three types: `getBy`, `queryBy` and `findBy`. Go through [the React Testing Library docs page about queries](https://testing-library.com/docs/queries/about/). Pay extra attention to the "Types of Queries" and "Priority" section. 
```

> As stated by the React Testing Library docs, `ByRole` methods are the favoured methods for querying, especially when paired with the `name` option. For example, we could improve the specificity of the above query like so: `getByRole("heading", { name: "Our First Test" })`. Queries through `ByRole` ensures that our UI is accessible to everyone no matter what mode they use to navigate the webpage ( i.e mouse or assistive technologies). 

Reviewer: ```suggestion
As stated by the React Testing Library docs, `ByRole` methods are the favored methods for querying, especially when paired with the `name` option. For example, we could improve the specificity of the above query like so: `getByRole("heading", { name: "Our First Test" })`. Queries that are done through `ByRole` ensure that our UI is accessible to everyone no matter what mode they use to navigate the webpage ( i.e mouse or assistive technologies). 
```

Snapshot tests may cause false positives. Since we cannot ascertain the validity of the component from a snapshot test, a bug might go undetected. Over-reliance on snapshots can make developers more confident about their code than they should be. 

The other issue with snapshots is false negatives. Even the most insignificant of changes compel the test to fail. Fixing punctuation? Snapshot will fail. Replacing an HTML tag to a more semantic one? Snapshot will fail. This might cause us to lose our confidence in the test suite altogether. Snapshots aren't inherently bad; they do serve a purpose. But it's beneficial to understand when to snapshot, and when not to snapshot. 

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Take a glance at all of the available [query methods](https://testing-library.com/docs/dom-testing-library/cheatsheet/). There's no need to use them all, but it's optimal to employ a specific method for a specific query. If none of the query methods suffice, there's an option to use [test ids](https://testing-library.com/docs/queries/bytestid/).

> 2. Read the [userEvent](https://testing-library.com/docs/ecosystem-user-event/) API docs to get a feel of how to achieve user simulation. 

Reviewer: ```suggestion
2. Read [the userEvent API docs](https://testing-library.com/docs/ecosystem-user-event/) to get a feel of how to achieve user simulation. 
```

### Introduction

We've learned testing our vanilla JavaScript applications in the previous section. Testing is indeed a powerful tool that allows to write maintainable and flexible code. If you've followed along our lessons, we've been using the [Jest](https://jestjs.io/) framework. For React, we'll keep using Jest and add more capabilities to our tests using the React Testing library.

### Learning Outcomes

By the end of this lesson, you should be able to do the following:

* Discern what packages will be needed to test a React app.
* Write queries to get elements from the UI and test them.
* Be able to simulate user events with the `userEvent` API.

> * Explain what a snapshot test is and how to carry out a snapshot test with Jest.

Reviewer: These aren't quite as high level as what I like to see. Rather than stating specifically what the student will be able to do our current direction for LOs is to outline the things that they'll learn about, i.e. "Test UI elements" vs "Write queries to get elements from the UI and test them."

In this [style guide update PR ](https://github.com/TheOdinProject/curriculum/pull/23051) we're moving towards having "Lesson Overviews" rather than even identifying them as "Learning Outcomes" to move towards this goal. Check the LO related content in that PR out and let me know if what I'm asking for makes sense.

Author: I've tried to make the LOs more of a general overview, though I'm still unsure with my changes. 

I've read through the style guide PR and I'm confused whether to change the "Learning Outcomes" heading to "Learning Overview". I've [commented on the style guide PR](https://github.com/TheOdinProject/curriculum/pull/23051#discussion_r791461163) for this matter.

3. This article on the [Pros and Cons of Jest Snapshot Tests](https://tsh.io/blog/pros-and-cons-of-jest-snapshot-tests/) goes in depth regarding the advantages and disadvantages of snapshot testing. And this one, [Snapshot Testing: Benefits and Drawbacks](https://www.sitepen.com/blog/snapshot-testing-benefits-and-drawbacks), does an excellent job of explaining what is snapshot testing in programming in general.

</div>

### Additional Resources

This section contains helpful links to other content. It isn't required, so consider it supplemental.

- This [tutorial on Testing React Apps by Academind](https://academind.com/tutorials/testing-react-apps) is a great overview of what you've learned. It goes into testing async code and callbacks which we haven't covered yet. Though you should be able to follow along using your previous knowledge.

- This [Intro to React Testing Library video](https://www.youtube.com/watch?v=YQLn7ycfzEo) for a hands-on tutorial.

### Knowledge Check 

This section contains questions for you to check your understanding of this lesson. If you're having trouble answering the questions below on your own, review the material above to find the answer.

> * <a class="knowledge-check-links" href="#setting-up">What packages are required for React testing?</a>

Reviewer: I'm not sure this would really be a core concept of the lesson, do you feel strongly that it is?

Author: It's imperative to understand how to set up a React test file. This entails that it's crucial to know what packages will be required to import into that file.

I don't argue with it being a core concept of the lesson, but I do believe that this particular knowledge check should stay.

* <a class="knowledge-check-links" href="#user-event">What is the significance of the user-event package?</a>
* <a class="knowledge-check-links" href="https://testing-library.com/docs/react-testing-library/api/#render">What does the `render` method do?</a>
* <a class="knowledge-check-links" href="#by-role-methods">What is the most preferred method for querying?</a>

> * <a class="knowledge-check-links" href="#simulating-user-events">How to test for a click event with `userEvent`?</a>

Reviewer: ```suggestion
* <a class="knowledge-check-links" href="#simulating-user-events">How would you test for a click event with `userEvent`?</a>
```

// App.test.js

import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";

describe("App component", () => {
  it("renders magnificent monkeys", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();

>   });

Reviewer: I'm curious why you elected to use restructuring from render here rather than using screen. I know that after our previous discussion you felt that one example destructuring from render should be kept in, but I assumed that was accomplished in lines 58-62?

Author: The main reason was to keep the `toMatchSnapshot` assertion here, since it ties pretty well with the next section "What are Snapshots?".

`container` is not a property on the `screen` object it seems, and this document confirms it: https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/prefer-screen-queries.md

It says: 

> DOM Testing Library (and other Testing Library frameworks built on top of it) exports a screen object which has every query (and a debug method). This works better with autocomplete and makes each test a little simpler to write and maintain.
> 
> This rule aims to force writing tests using built-in queries directly from screen object rather than destructuring them from render result. Given the screen component does not expose utility methods such as rerender() or the container property, it is correct to use the render returned value in those scenarios.

One way I could achieve the snapshot test without using `container` was using `document`, but I don't know if that's ideal:

```js
  it("renders magnificent monkeys", () => {
    render(<App />);
    expect(document.body.firstChild).toMatchSnapshot();
  });
```

Otherwise, an alternate way to test would be to just use `screen.getByRole` to assert that the heading is there. That way, we'll not need the `container` at all.

For the time being, I've added a comment to justify render's destructuring in this test:

```js
  it("renders magnificent monkeys", () => {
    // since screen does not have the container property, we'll destructure render to obtain container for this test
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
 ```
 
 Let me know if we should remove/edit this comment. Also let me know if we want to keep this test as is or go with a different approach.

  it("renders radical rhinos after button click", () => {
    render(<App />);
    const button = screen.getByRole("button", { name: "Click Me" });

    userEvent.click(button);

    expect(screen.getByRole("heading").textContent).toMatch(/radical rhinos/i);
  });
});
~~~

The tests speak for themselves. In the first test, we utilize snapshots to check whether all the nodes render as we expect them to. In the second test, we simulate a click event. Then we check if the heading changed. `toMatch` is one of the various assertions we could have made. 

> What's this `screen` object? `screen` is another utility that React Testing Library provides. It has all the methods for querying. With `screen`, we don't have to worry about keeping `render`'s destructuring up-to-date. Hence, it's better to use `screen` to access queries rather than to destructure `render`.

Reviewer: I would move the `screen` description above the examples using it.
